File: programming_language/javasolutions/p84/Solution.py
# http://www.cnblogs.com/lichen782/p/leetcode_Largest_Rectangle_in_Histogram.html

class Solution:
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        if not heights:
            return 0

        asc_height_idx_stack = []
        max_area = 0
        size = len(heights)
        for i in range(size + 1):
            current_height = heights[i] if i < size else -1

            while asc_height_idx_stack and current_height <= heights[asc_height_idx_stack[-1]]:
                top = asc_height_idx_stack.pop()

                if asc_height_idx_stack:
                    area_with_top = heights[top] * (i - asc_height_idx_stack[-1] - 1)
                else:
                    # why the hell the width is i
                    # If asc_height_idx_stack is empty, that means
                    # heights[top] is the "lowest" bar before heights[i] bar
                    # from 0 to i (excluding i), the width is i + 1 - 1
                    area_with_top = heights[top] * i
                max_area = max(max_area, area_with_top)
            asc_height_idx_stack.append(i)
        
        return max_area

def main():
    sol = Solution()
    print(sol.largestRectangleArea([2, 1, 5, 6, 2, 3]))

if __name__ == '__main__':
    main()

File: programming_language/javasolutions/p84/Solution.java
/**
 * Given n non-negative integers representing the histogram's bar height where
 * the width of each bar is 1, find the area of largest rectangle in the histogram.
 */

package javasolutions.p84;

import java.util.Stack;

public class Solution {
  public int largestRectangleArea(int[] heights) {
    if(heights.length == 0) return 0;

    int answer = 0;
    for(int i = 0; i < heights.length; i++) {
      if(i + 1 < heights.length && heights[i] <= heights[i+1]) {
        continue;
      }

      int minHeight = heights[i];
      for(int j = i; j >= 0; j--) {
        minHeight = Math.min(minHeight, heights[j]);
        int currentArea = minHeight * (i - j + 1);
        answer = Math.max(answer, currentArea);
      }
    }

    return answer;
  }

  /**
   * http://www.geeksforgeeks.org/largest-rectangle-under-histogram/
   * For every height `x`, we calculate the area with `x` as the smallest bar
   * in the rectangle.
   * If we calculate such area with every height and find the max of all areas,
   * then we find the answer.
   */
  public int largestRectangleAreaWithStack(int[] heights) {
    if(heights.length == 0) return 0;

    // Create an empty stack. The stack holds indexes of hist[] array
    // The bars stored in stack are always in increasing order of their
    // heights.
    Stack<Integer> stack = new Stack<>();
    int top;
    int areaWithTop;
    int maxArea = 0;

    int i = 0;
    while(i < heights.length) {
      if(stack.isEmpty() || heights[i] > heights[stack.peek()]) {
        stack.push(i);
        i++;
      } else {
        top = stack.pop();
        // now height[i] < height[top], height[top] >= height[stack.peek()]
        if(stack.isEmpty()) {
          areaWithTop = heights[top] * i;
        } else {
          areaWithTop = heights[top] * (i - stack.peek() - 1);
        }

        maxArea = Math.max(maxArea, areaWithTop);
      }
    }

    // i = heights.length after the above loop, stack could be non-empty
    while(!stack.isEmpty()) {
      top = stack.pop();

      if(stack.isEmpty()) {
        areaWithTop = heights[top] * i;
      } else {
        areaWithTop = heights[top] * (i - stack.peek() - 1);
      }

      maxArea = Math.max(maxArea, areaWithTop);
    }

    return maxArea;
  }
}
